File: 2024/jl/day01.jl
# Day 1: Historian Hysteria

include("util.jl")

using .Util

lines = Util.getinput("day01.txt")

function star1()
	id1 = []
	id2 = []

	for line ∈ lines
		if line == ""
			break
		end

		a, b = split(line, "   ")
		push!(id1, parse(Int64, a))
		push!(id2, parse(Int64, b))
	end

	id1 = sort(id1)
	id2 = sort(id2)

	result = sum([abs(x - y) for (x, y) ∈ zip(id1, id2)])
	println("- $result")
end

function star2()
	llist = []
	summap = Dict()

	for line ∈ lines
		if line == ""
			break
		end

		a, b = split(line, "   ")
		
		a = parse(Int64, a)
		push!(llist, a)

		b = parse(Int64, b)
		if !haskey(summap, b)
			summap[b] = 1
		else
			summap[b] += 1
		end
	end

	result = sum([(n * get(summap, n, 0)) for n ∈ llist])
	println("- $result")
end

star1()
star2()

File: 2024/py/src/day01.py
# Day 1: Historian Hysteria

from src import util


def sep():
    txt = util.input_lines("day01.txt")

    id1, id2 = [], []
    for line in txt:
        if not line:
            break

        a, b = line.strip().split("   ")
        id1.append(int(a))
        id2.append(int(b))

    return id1, id2


def star1():
    id1, id2 = sep()
    id1.sort()
    id2.sort()

    sum = 0
    for a, b in zip(id1, id2):
        sum += abs(a - b)

    return sum


def star2():
    id1, id2 = sep()

    sum = 0
    for a in id1:
        sum += a * id2.count(a)

    return sum


util.run_day(1, star1, star2)
